### DISCUSSION TITLE
React Native support

### DISCUSSION THREAD

**Discussion by Cancuuu** [wevm/wagmi#533] [Q&A]
Hey guys, how are you? just a little question: 
wagmi.sh works for React Native or Expo ? 

Thanks!

**Comment by wobsoriano** [comment]
RainbowKit is React Native

EDIT: I meant the rainbow mobile app

**Reply by Cancuuu** [reply]
Are you sure that you can use https://github.com/rainbow-me/rainbowkit  in a RN project?

**Reply by wobsoriano** [reply]
IIRC https://rainbow.me/ is made with RainbowKit and RN

**Reply by Cancuuu** [reply]
https://github.com/rainbow-me/rainbow/blob/develop/package.json

**Reply by wobsoriano** [reply]
Looks like they went with vanilla ethers + react-query. You should still be able to use wagmi  with RN tho since it has no dom functionality

**Reply by wobsoriano** [reply]
Just add the polyfills too - `npm install browserify-zlib buffer  util events` in your RN app

Also check this https://github.com/rainbow-me/rainbow/blob/develop/shim.js

**Reply by Cancuuu** [reply]
Ill try it! thank you so much man! 

If you wanna talk add me on Telegram or Twitter: 
https://t.me/cancu7
https://twitter.com/CancuDev

**Reply by peterferguson** [reply]
Hey I started working on this last weekend you need to create a custom connector for `WalletConnectConnector` to work,

I was able able to hack one together last weekend but I am planning on making a pr here and seeing if it can be added.

On top of the custom connector your either need to create your own wallet connect modal UI or add in the official wallet connect react native modal & hack your connector to open it instead of the standard web version

**Reply by Cancuuu** [reply]
Excellent! so you say that wagmi.sh works for React Native, but the WalletConnectConnector doesnt? Are you trying make a PR adding a WalletConnectConnector for RN, here in Wagmi.sh?

**Reply by wobsoriano** [reply]
Yes wagmi + polyfills + custom wallet connector

**Reply by wobsoriano** [reply]
you can hack `@walletconnect/react-native-dapp` too and use their own connect function then pass the connector to walletconnector... something like

```ts
import { useWalletConnect } from '@walletconnect/react-native-dapp';
import { WalletConnectConnector } from 'wagmi/connectors/walletConnect'
import { useConnect } from 'wagmi'

function Connect() {
  const connector = useWalletConnect();
  
  const { connect } = useConnect({
    connector: new WalletConnectConnector({
      options: {
        qrcode: false,
        connector
      }
    })
  })

  useEffect(() => {
    if (connector?.accounts?.length) {
      connect()
    }
  }, [connector])

 // <Button title="Connect" onPress={() => connector?.connect()} />
```

**Reply by wobsoriano** [reply]
Most of wagmi hooks use a custom [`useQuery`](https://github.com/tmm/wagmi/blob/main/packages/react/src/hooks/utils/useQuery.ts) implementation that stops it from fetching in React Native

**Reply by peterferguson** [reply]
Yea this is pretty much the set up I had but I hadn't got round to testing too many the hooks just the connecting wallet part and using `useAccount`

**Reply by peterferguson** [reply]
> Excellent! so you say that wagmi.sh works for React Native, but the WalletConnectConnector doesnt? Are you trying make a PR adding a WalletConnectConnector for RN, here in Wagmi.sh?

I had planned to and reached out to the authors no twitter, they were up for it but just haven't had time to get round to it yet

**Reply by peterferguson** [reply]
Is the custom `useQuery` implementation failing because of the `storage` implemenation though? 

By default it uses `localStorage` when I implemented it I used `createStorage` and [`react-native-mmkv`](https://github.com/mrousavy/react-native-mmkv) as the storage and the hooks I used worked.

@wobsoriano Did you dig any deeper as to why the custom useQuery didn't work?

**Reply by wobsoriano** [reply]
@peterferguson wagmi exports their custom noopStorage for temporary making the storage work but even with that the query is still not running

```ts
const client = createClient({
  persister: asyncStoragePersistor,
  storage: createStorage({
    storage: noopStorage
  })
})
```

**Reply by wobsoriano** [reply]
@peterferguson  I tried copy-pasting their `useQuery` implementation so I can play with it better but can't seem to find the problem

**Reply by jxom** [reply]
There is nothing platform specific in `useQuery`, so it should work in a React Native environment given you provide your own `storage` as @peterferguson suggests, or none at all.

**Reply by wobsoriano** [reply]
If you don’t provide any storage it will look for window.localStorage

**Reply by jxom** [reply]
By default, it will use `window.localStorage` yeah. Using `noopStorage` is the equivalent of providing no storage.

**Reply by wobsoriano** [reply]
Yes that’s what I did in the example I commented… just curious why the fetching isnt working still

**Reply by tmm** [reply]
Maybe related to `useSyncExternalStore`. Looks like we need to switch between the plain and native shims.

**Reply by wobsoriano** [reply]
@tmm, I tried commenting out `useSyncExternalStore` and that's indeed the issue.

**Reply by wobsoriano** [reply]
funny this works...

original: `import { useSyncExternalStore } from 'use-sync-external-store/shim/index'`

patched: `import { useSyncExternalStore } from 'use-sync-external-store/shim'`

without the `index` it's now fetching... [patch-package](https://github.com/ds300/patch-package) to the rescue.

I'll post a repo here later

**Reply by peterferguson** [reply]
yep can confirm removing the fully qualified shim import makes it work

**Reply by peterferguson** [reply]
> RainbowKit is React Native
> 
> EDIT: I meant the rainbow mobile app

@jxom Question now is WEN RainbowKit-RN lol

**Reply by thegitparticle** [reply]
The new updated wage 0.64.0 is throwing this error - https://github.com/wagmi-dev/wagmi/issues/911
The build works fine, but on open, the app throws getProvider() can't find.

Anybody who used the latest Wagmi version in react native? Did it work properly or what workaround did you try?

**Comment by wobsoriano** [comment]
Here's the working example with WalletConnect modal 

https://github.com/wobsoriano/wagmi-react-native-starter

**Reply by thegitparticle** [reply]
The new updated wage 0.64.0 is throwing this error - [#911](https://github.com/wagmi-dev/wagmi/issues/911) 
The build works fine, but on open, the app throws getProvider() can't find.

**Reply by felipecamposfabel** [reply]
@wobsoriano Thanks for the example App. It helped me a lot to make my app works with WalletConnect. I also tried to make some workarounds to get Coinbase to work with wagmi on RN, but with no success. 

I tried working with the mobile version of their SDK and creating a custom connector... https://github.com/coinbase/wallet-mobile-sdk/blob/master/react-native/client/example/App.js

Do you know any way to achieve that?

**Reply by thegitparticle** [reply]
@felipecamposfabel do you have repo of wagmi and walletconnect working with the latest versions?

**Reply by felipecamposfabel** [reply]
@thegitparticle Right now it's on a private repo. But with the example wobsoriano created, it works. Pay attention to the `package.json`, `shim.js` and `metro.config`. For the latest version, you don't need the patches they mentioned. Also, I would recommend using zustand as a storage, if you already use it on your project... Something like this, should work as a storage, instead of using a persister with noopStorage...

```export const useWeb3 = create(
  persist(
    (set, get) => ({
      getItem: (key, defaultState) => get()?.[key] ?? defaultState,
      setItem: (key, value) => set({ [key]: value }),
      removeItem: (key) => set((state) => _.omit(state, key), true),
    }),
    {
      name: 'web3',
      storage: createJSONStorage(() => AsyncStorage),
    }
  )
);

export const wagmiStorage = {
  getItem: useWeb3.getState().getItem,
  setItem: useWeb3.getState().setItem,
  removeItem: useWeb3.getState().removeItem,
};
```
I also had some issues with react-native-svg. Make sure to reinstall your app.

**Reply by thegitparticle** [reply]
thanks, I'm gonna try this and update.

But, is this for WalletConnect 2.0 or v1?
